paasta_tools/deployd/watchers.py

```diff
@@ -45,6 +55,135 @@ def startup_checker(self, obj):
+                service_instances.append(ServiceInstance(service=service,
+                                                         instance=instance,
+                                                         bounce_by=int(time.time()),
+                                                         watcher=self.__class__.__name__,
+                                                         bounce_timers=None))
+        return service_instances
+
+
+class PublicConfigEventHandler(pyinotify.ProcessEvent):
+
+    def my_init(self, filewatcher):
+        self.filewatcher = filewatcher
+        self.public_config = load_system_paasta_config()
+
+    @property
+    def log(self):
+        name = '.'.join([__name__, self.__class__.__name__])
+        return logging.getLogger(name)
+
+    def filter_event(self, event):
+        if event.name.endswith('.json'):
+            return event
+
+    def watch_new_folder(self, event):
+        if event.maskname == 'IN_CREATE|IN_ISDIR':
+            self.filewatcher.wm.add_watch(event.pathname, self.filewatcher.mask, rec=True)
+
+    def process_default(self, event):
+        self.log.debug(event)
+        self.watch_new_folder(event)
+        event = self.filter_event(event)
+        if event:
+            self.log.debug("Public config changed on disk, loading new config")
+            try:
+                new_config = load_system_paasta_config()
+            except ValueError:
+                self.log.error("Couldn't load public config, the JSON is invalid!")
+                return
+            service_instances = []
+            if new_config != self.public_config:
+                self.log.info("Public config has changed, now checking if it affects any services config shas")
+                self.public_config = new_config
+                service_instances = self.get_service_instances_with_changed_id()
+            if service_instances:
+                self.log.info("Found config change affecting {} service instances, "
+                              "now doing a staggered bounce".format(len(service_instances)))
+                bounce_rate = self.public_config.get_deployd_big_bounce_rate()
+                service_instances = rate_limit_instances(instances=service_instances,
+                                                         number_per_minute=bounce_rate,
+                                                         watcher_name=self.__class__.__name__)
+            for service_instance in service_instances:
+                self.filewatcher.inbox_q.put(service_instance)
+
+    def get_service_instances_with_changed_id(self):
+        marathon_app_ids = [app.id.lstrip('/') for app in get_marathon_apps()]
+        instances = get_services_for_cluster(cluster=self.public_config.get_cluster(),
+                                             instance_type='marathon',
+                                             soa_dir=DEFAULT_SOA_DIR)
+        service_instances = []
+        for service, instance in instances:
```

Comment:
This sorta overlaps with this
https://github.com/Yelp/paasta/blob/master/paasta_tools/list_marathon_service_instances.py#L91

this is a pretty expensive operation, luckily it will only happen on public config changes